### PR TITLE
Refactor workspaces

### PR DESCRIPTION
**Pull request by eseliger**

### CODE REVIEW

File: internal/batches/errors.go [hunk 1]
```diff
@@ -77,3 +77,9 @@ func (e IgnoredRepoSet) Append(repo *graphql.Repository) {
 func (e IgnoredRepoSet) HasIgnored() bool {
 	return len(e) > 0
 }
+
+type ValidationError struct{ Reason string }
+
+func (e ValidationError) Error() string {
+	return fmt.Sprintf("validation failed: %s", e.Reason)
+}
```
Comment (eseliger) [inline internal/batches/errors.go:85]: I think I want to have this to distinguish validation things from actual errors, so we can in the future have better error reporting. Can remove it, though.

File: internal/batches/service/task_builder.go [hunk 2]
```diff
@@ -81,17 +80,24 @@ func (tb *taskBuilder) buildTask(r *graphql.Repository, path string, onlyWorkspa
...
 
-	var tasks []*executor.Task
-	for repo, ws := range workspaces {
-		for _, path := range ws.paths {
-			t, ok, err := tb.buildTask(repo, path, ws.onlyFetchWorkspace)
+	tasks := []*executor.Task{}
+	for _, ws := range workspaces {
+		repo, ok := repoByID[ws.RepoID]
+		if !ok {
+			return nil, errors.New("invalid state")
```
Comment (eseliger) [inline internal/batches/service/task_builder.go:93]: it is, right?

File: internal/batches/service/workspaces.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+// workspaces. root contains the repositories that didn't match a config.
+// If the user didn't specify any workspaces, the repositories are returned as
+// root repositories.
+func findWorkspaces(
+	ctx context.Context,
+	spec *batcheslib.BatchSpec,
+	finder directoryFinder,
+	repos []*graphql.Repository,
+) ([]RepoWorkspaces, error) {
+	// Pre-compile all globs.
```
Comment (eseliger) [inline internal/batches/service/workspaces.go:32]: This function and type can relatively easy be pulled over to the backend now. Next step :)

File: internal/batches/service/workspaces.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
+package service
```
Comment (mrnugget) [inline internal/batches/service/workspaces.go:1]: I think we need to move some tests too

File: internal/batches/service/workspaces.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	RepoID             string
+	Paths              []string
+	OnlyFetchWorkspace bool
+}
+
+// findWorkspaces matches the given repos to the workspace configs and
+// searches, via the Sourcegraph instance, the locations of the workspaces in
+// each repository.
+// The repositories that were matched by a workspace config are returned in
+// workspaces. root contains the repositories that didn't match a config.
```
Comment (mrnugget) [inline internal/batches/service/workspaces.go:23]: Comment is outdated

**Review by mrnugget — APPROVED** [review]

File: internal/batches/service/workspaces.go (hunk 4 above)
Comment (eseliger) [reply internal/batches/service/workspaces.go:1]: Done!

**Review by mrnugget — APPROVED** [review]

**Pull request merged**